cmd/schema-tweak/main.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2024 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const resourcesDir = "config/core/300-resources"
+
+func main() {
+	files, err := os.ReadDir(resourcesDir)
+	if err != nil {
+		log.Fatalln("failed to read CRD directory", err)
+	}
+
+	for _, file := range files {
+		if !strings.HasSuffix(file.Name(), ".yaml") || !file.Type().IsRegular() {
+			continue
+		}
+
+		processFile(file)
+	}
+
+}
+
+func processFile(file fs.DirEntry) {
+	fmt.Println("Processing", file.Name())
+
+	filename := filepath.Join(resourcesDir, file.Name())
+
+	fbytes, err := os.ReadFile(filename)
+	if err != nil {
+		log.Fatalln("failed to read CRD", err)
+	}
+
+	var root yaml.Node
+	if err := yaml.Unmarshal(fbytes, &root); err != nil {
+		log.Fatalln("failed to unmarshal CRD", err)
+	}
+
+	buf := bytes.Buffer{}
+
+	enc := yaml.NewEncoder(&buf)
+	enc.SetIndent(2)
+
+	// document node => content[0] => crd map
+	applyOverrides(root.Content[0])
+
+	if err = enc.Encode(&root); err != nil {
+		log.Fatalln("failed to marshal CRD", err)
+	}
+
+	if err = enc.Close(); err != nil {
+		log.Fatalln("failed to close yaml encoder", err)
+	}
+
+	if err = os.WriteFile(filename, buf.Bytes(), file.Type().Perm()); err != nil {
+		log.Fatalln("failed to write CRD", err)
+	}
+
+}
+
+func applyOverrides(root *yaml.Node) {
+	for _, override := range overrides {
+		crdName := stringValue(root, "metadata.name")
+
+		if crdName != override.crdName {
+			continue
+		}
+
+		versions := arrayValue(root, "spec.versions")
+
+		for _, version := range versions {
+			for _, entry := range override.entries {
+				schemaNode := mapValue(version, "schema.openAPIV3Schema")
+				applyOverrideEntry(schemaNode, entry)
+			}
+		}
+	}
+}
+
+func applyOverrideEntry(node *yaml.Node, entry entry) {
+	for _, segment := range strings.Split(entry.path, ".") {
```

Comment:
I guess there is no way to quickly reference a path (its node) so we have to iterate.